Add missing docs for release 12.2.0 (#3582)
PostgREST/postgrest: docs/references/listener.rst

.. _listener:

Listener
########

> PostgREST uses PostgreSQL `LISTEN <https://www.postgresql.org/docs/current/sql-listen.html>`_ to reload its :ref:`Schema Cache <schema_reloading_notify>` and :ref:`Configuration <config_reloading_notify>` via `NOTIFY <https://www.postgresql.org/docs/current/sql-notify.html>`_. This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals. Like on cloud managed containers or on Windows systems.

Reviewer: ```suggestion
PostgREST uses `LISTEN <https://www.postgresql.org/docs/current/sql-listen.html>`_ to reload its :ref:`Schema Cache <schema_reloading_notify>` and :ref:`Configuration <config_reloading_notify>` via `NOTIFY <https://www.postgresql.org/docs/current/sql-notify.html>`_. This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals. Like on cloud managed containers or on Windows systems.
```

.. _listener:

Listener
########

PostgREST uses `LISTEN <https://www.postgresql.org/docs/current/sql-listen.html>`_ to reload its :ref:`Schema Cache <schema_reloading_notify>` and :ref:`Configuration <config_reloading_notify>` via `NOTIFY <https://www.postgresql.org/docs/current/sql-notify.html>`_.
This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals.
Like on cloud managed containers or on Windows systems.

.. code:: postgresql

  NOTIFY pgrst, 'reload schema'; -- reload schema cache
  NOTIFY pgrst, 'reload config'; -- reload config
  NOTIFY pgrst;                  -- reload both

By default, the LISTEN channel is enabled (:ref:`db-channel-enabled`) and named ``pgrst`` (:ref:`db-channel`).

Listener on Read Replicas
=========================

The ``LISTEN`` and ``NOTIFY`` commands do not work on PostgreSQL read replicas.
Thus, if you connect PostgREST to a read replica the Listener will fail to start.

.. code:: console

>   12/Jun/2024:15:19:57 -0500: Failed listening for notifications on the "pgrst" channel... connection to server... failed: session is read-only

Reviewer: Ah, damn. I just realized this error doesn't make sense here, it's caused because of our `target_session_attrs=read-write`.

The previous one with plain SQL was much better. Maybe add something like:

```postgresql
-- check if the instance is a replica
postgres=# select pg_is_in_recovery();
 pg_is_in_recovery 
-------------------
 t
 (1 row)

postgres=# LISTEN pgrst;
<ERROR message here>
```

> To work around this, you can connect the Listener to a primary cluster while still using the :ref:`connection_pool` on the replica.

Reviewer: ```suggestion
To work around this, you can connect the Listener to a primary while still using the :ref:`connection_pool` on the replica.
```

.. _listener:

Listener
########

PostgREST uses PostgreSQL `LISTEN <https://www.postgresql.org/docs/current/sql-listen.html>`_ to reload its :ref:`Schema Cache <schema_reloading_notify>` and :ref:`Configuration <config_reloading_notify>` via `NOTIFY <https://www.postgresql.org/docs/current/sql-notify.html>`_.
This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals.
Like on cloud managed containers or on Windows systems.

.. code:: postgresql

  NOTIFY pgrst, 'reload schema'; -- reload schema cache
  NOTIFY pgrst, 'reload config'; -- reload config
  NOTIFY pgrst;                  -- reload both

By default, the LISTEN channel is enabled (:ref:`db-channel-enabled`) and named ``pgrst`` (:ref:`db-channel`).

Listener on Read Replicas
=========================

The ``LISTEN`` and ``NOTIFY`` commands do not work on PostgreSQL read replicas.
Thus, if you connect PostgREST to a read replica the Listener will fail to start.

.. code:: postgresql

  NOTIFY pgrst;

>   ERROR:  cannot execute NOTIFY during recovery

Reviewer: I think we should put the actual PostgREST error here. IIRC it says the same but with additional logs.

Something like this should work:

```
$ postgrest-with-* --replica postgrest-run
```

Maybe you'll have to edit the withTmp on the nix file

Author: The error in `psql` is exactly the same:

```
postgres=# notify pgrst;
ERROR:  cannot execute NOTIFY during recovery
```

Now, in the PostgREST 12.2.0 logs, when connecting to a read replica only, it shows:

```
12/Jun/2024:15:19:57 -0500: Failed listening for notifications on the "pgrst" channel. Just "connection to server on socket \"/run/user/1000/postgrest/postgrest-with-postgresql-16-Y3M/socket_replica_6039/.s.PGSQL.5432\" failed: session is read-only\n"
12/Jun/2024:15:19:57 -0500: Failed listening for notifications on the "pgrst" channel. ExitFailure 1
12/Jun/2024:15:19:57 -0500: Attempting to connect to the database...
12/Jun/2024:15:19:57 -0500: Retrying listening for notifications in 8 seconds...
```

Are you referring to the latter?

Reviewer: Yeah, maybe just put:

```
12/Jun/2024:15:19:57 -0500: Failed listening for notifications on the "pgrst" channel... connection to server .. failed: session is read-only
```

That error message needs fixing btw. Can be another issue.

> To work around this, you can connect the Listener to a primary database while still using the :ref:`connection_pool` on the replica.

Reviewer: ```suggestion
To work around this, you can connect the Listener to a primary cluster while still using the :ref:`connection_pool` on the replica.
```

.. _listener:

Listener
########

> PostgREST enables a `listener to a notification channel <https://www.postgresql.org/docs/current/sql-listen.html>`_ at startup.
> It's used to :ref:`reload the Schema Cache <schema_reloading_notify>` and the :ref:`Configuration <config_reloading_notify>` from within the database via the `NOTIFY command <https://www.postgresql.org/docs/current/sql-notify.html>`_.
> 
> .. code:: postgresql
> 
>   NOTIFY pgrst;
> 
> The ``pgrst`` notification channel is enabled by default.
> See :ref:`db-channel` and :ref:`db-channel-enabled` to configure this channel.
> 
> This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals.
> Like on cloud managed containers or on Windows systems.

Reviewer: Compressing a bit and adding the accepted messages

```suggestion
PostgREST uses PostgreSQL `LISTEN <https://www.postgresql.org/docs/current/sql-listen.html>`_ to reload its :ref:`Schema Cache <schema_reloading_notify>` and :ref:`Configuration <config_reloading_notify>` via `NOTIFY <https://www.postgresql.org/docs/current/sql-notify.html>`_. This is useful in environments where you can’t send SIGUSR1 or SIGUSR2 Unix Signals. Like on cloud managed containers or on Windows systems.

.. code:: postgresql

  NOTIFY pgrst, 'reload schema'; -- reload schema cache
  NOTIFY pgrst, 'reload config'; -- reload config
  NOTIFY pgrst;                  -- reload both

By default, the LISTEN channel is enabled (:ref:`db-channel-enabled`) and named `"pgrst"` (:ref:`db-channel`).
```

Listener on Read Replicas
=========================

> The ``LISTEN`` and ``NOTIFY`` commands do not work on PostgreSQL read replicas.
> 
> .. code:: postgresql
> 
>   NOTIFY pgrst;
> 
> .. code::
> 
>   ERROR:  cannot execute NOTIFY during recovery

Reviewer: ```suggestion

The ``LISTEN`` command does not work on PostgreSQL read replicas.

.. code:: postgresql

  NOTIFY pgrst;
  ERROR:  cannot execute NOTIFY during recovery
```

> If PostgREST connects to a read replica, the notification channel will fail to start.
> To prevent this, you need to provide `two hosts <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-MULTIPLE-HOSTS>`_ in your :ref:`connection string <db-uri>`.
> The first one would be the read replica and the second one the primary database.
> For example:
> 
> .. code:: bash
> 
>   db-uri = "postgres://read_replica:5433,primary:5432/mydb"
> 
> This way, PostgREST will handle requests in the ``read_replica`` host and listen to notifications in the fallback ``primary`` host.
> 
> .. note::
> 
>   Under the hood, PostgREST adds the parameter `target_session_attrs=read-write <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-TARGET-SESSION-ATTRS>`_ to the connection string when creating a listener session in the DB.

Reviewer: ```suggestion
To workaround this, you can connect the Listener to a primary database while still using the :ref:`connection_pool` on the replica.

To do this, use the standard `libpq multiple hosts <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-MULTIPLE-HOSTS>`_ and `target_session_attrs <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-TARGET-SESSION-ATTRS>`_ in your :ref:`connection string <db-uri>`.

.. code:: bash

  db-uri = "postgres://read_replica.host,primary.host/mydb?target_session_attrs=read-only"

This will cause the :ref:`connection_pool` to connect to the read replica host and LISTEN on the fallback primary host.

.. note::

  Under the hood, PostgREST forces `target_session_attrs=read-write <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-TARGET-SESSION-ATTRS>`_ for the LISTEN session.
```
